index.bs

```diff
@@ -12337,18 +12336,77 @@ The value of the [=function object=]’s <code class="idl">name</code> property
 
 <h3 id="es-platform-objects">Platform objects implementing interfaces</h3>
 
+<div algorithm>
+    An ECMAScript value |value| <dfn id="es-is-platform-object">is a platform object</dfn>
+    if [$Type$](|value|) is Object and if |value| has a \[[PrimaryInterface]] internal slot.
+</div>
+
+<div algorithm>
+    An ECMAScript value |value| <dfn export>implements</dfn> an [=interface=] |interface|
+    if |value| [=is a platform object=] and one of the following is true:
+    <ul>
+        <li>|value|.\[[PrimaryInterface]] is |interface|.</li>
+        <li>The set of [=inherited interfaces=] of |value|.\[[PrimaryInterface]] contains |interface|.</li>
+    </ul>
+</div>
+
+<div class="note">
+    Specifications use various wordings to reference the "[=implements=]" algorithm,
+    such as "|object| is an |interface| object", to mean, "|object| implements |interface|".
+</div>
+
 Every [=platform object=] is associated with a global environment, just
 as the [=initial objects=] are.
 It is the responsibility of specifications using Web IDL to state
 which global environment (or, by proxy, which global object) each platform
 object is associated with.
 
+<div algorithm>
+    To <dfn>create an object implementing the interface</dfn> |interface|, with optional
+    intenal slots |slots|, for an interface which is not declared with the
+    [{{Global}}] [=extended attribute=], perform the following steps:
+
+    1. If |slots| is provided, append \[[PrimaryInterface]] to |slots|.
+    1. Otherwise, let |slots| be &laquo; \[[PrimaryInterface]] &raquo;.
+    1. Let |proto| be the [=interface prototype object=] of |interface|.
+    1. Let |instance| be [$ObjectCreate$](|proto|, |slots|).
+    1. Set |instance|.\[[PrimaryInterface]] to |interface|.
+    1. Let |realm| be the current [=Realm=].
```

Comment:
What if there isn't one?  Lots of objects get created when there is no current Realm around.  As a simple example, documents are (and this matters, because they have unforgeable things on them).  Same thing for events, actually.

This really needs to use the Realm of the passed-in global, not the current Realm.

Reply:
FWIW, I've been hoping that for cases where we create something before we create a realm, we can move those around, at least specification-wise. It seems better to first create a realm and global, and then a document.

Reply:
Oh, we may have a Realm fine.  But it's not going to be current unless either script is running or someone takes steps to make it current.

Reply:
I can see how this is an important issue, but I am not sure how to address it in this patch. If I add a realm argument which is required for cases where there is no current realm, that won't make everyone pass it.

Reply:
In previous discussions we wanted to default to the relevant realm, not the current one. And indeed it should be required in some scenarios; it would be a spec bug not to pass it. https://github.com/heycam/webidl/issues/135

Reply:
> that won't make everyone pass it.

True, but that way we can work on fixing those places to pass it, or have some default convention for what Realm is used when calling this algorithm in various situations.

Reply:
It seems like we have two modes where this algorithm will be called:
- Passing a new.target value (from all constructors)
- Somehow indicating a realm. I made this PR vaguely refer to the relevant realm of the relevant *this* value, but I'm not sure how we should formalize this or let folks pass in their own *this*/realm.

Could we say that, if you don't have a relevant *this* value, then you should explicitly pass in the new.target value? Specs could use wording like this:

> Create a new object implementing the XYZ interface, with new.target of the interface prototype object of the relevant realm of abc.

Or is that too much cryptic boilerplate?

Reply:
Too much cryptic boilerplate. As I mentioned above, I think the usage should be one of

> Create a new X object

or, if there is no this value

> Create a new X object in the relevant realm of abc